### PR TITLE
8189282: JavaFX: Invalid position of candidate pop-up of InputMethod in Hi-DPI on Windows

### PR DESCRIPTION
**Pull request by beldenfox**
When reporting input method candidate position the code in GlassViewEventHandler is not applying the platform scale factors. This is causing incorrect IM positions to be reported to glass on hi-dpi monitors.

This PR a no-op on Mac since the platform scale factors are always 1.0. I don't think it affects the current Linux XIM code at all but XIM is so out-of-date it has become difficult to test. PR #1080 will replace the old XIM code and needs this fix to work properly on hi-dpi screens.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8189282](https://bugs.openjdk.org/browse/JDK-8189282): JavaFX: Invalid position of candidate pop-up of InputMethod in Hi-DPI on Windows (**Bug** - P3)
 * [JDK-8254126](https://bugs.openjdk.org/browse/JDK-8254126): the position of Chinese Input Method candidates window is wrong (**Bug** - P3)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1311/head:pull/1311` \
`$ git checkout pull/1311`

Update a local copy of the PR: \
`$ git checkout pull/1311` \
`$ git pull https://git.openjdk.org/jfx.git pull/1311/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1311`

View PR using the GUI difftool: \
`$ git pr show -t 1311`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1311.diff">https://git.openjdk.org/jfx/pull/1311.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1311#issuecomment-1869661792)

### CODE REVIEW

File: .idea/misc.xml [hunk 1]
```diff
@@ -1,4 +1,3 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kevinrushforth) [inline .idea/misc.xml:1]: The changes in this file should be reverted, since it isn't related to the bug being fixed.

**Review by kevinrushforth — COMMENTED** [review]
The fix looks correct to me. I'll test it on Windows as part of formally reviewing it.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassViewEventHandler.java [hunk 2]
```diff
@@ -669,9 +669,12 @@ protected void onChanged(ListChangeListener.Change<InputMethodTextRun> c) {
     @Override
     public double[] getInputMethodCandidatePos(int offset) {
         Point2D p2d = scene.inputMethodRequests.getTextLocation(offset);
+        final Screen s = Screen.getMainScreen();
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassViewEventHandler.java:672]: This won't work for multiple screens. You'll need something like `scene.getPlatformView().getWindow().getScreen()` (with the appropriate null checks).

File: modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassViewEventHandler.java [hunk 3]
```diff
@@ -669,9 +669,12 @@ protected void onChanged(ListChangeListener.Change<InputMethodTextRun> c) {
...
+        float pScaleX = (s == null) ? 1.0f : s.getPlatformScaleX();
+        float pScaleY = (s == null) ? 1.0f : s.getPlatformScaleY();
         double[] ret = new double[2];
-        ret[0] = p2d.getX();
-        ret[1] = p2d.getY();
+        ret[0] = p2d.getX() * pScaleX;
+        ret[1] = p2d.getY() * pScaleY;
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassViewEventHandler.java:677]: You should use `Screen::toPlatformX/Y` or it won't work for all screen arrangements.

**Review by kevinrushforth — COMMENTED** [review]
This works fine for a single screen, but not for multiple screens. Also, it isn't sufficient to multiply the x and y by the platform scale for screens that don't touch the origin of the virtual screen space. See [this comment in PR #853](https://github.com/openjdk/jfx/pull/853#issuecomment-1206747079). I left inline comments with what I think needs to be changed.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassViewEventHandler.java [hunk 4]
```diff
@@ -672,6 +673,16 @@ public double[] getInputMethodCandidatePos(int offset) {
         double[] ret = new double[2];
         ret[0] = p2d.getX();
         ret[1] = p2d.getY();
+
+        for (Screen scr : Screen.getScreens()) {
+            Rectangle2D bounds = new Rectangle2D(scr.getX(), scr.getY(), scr.getWidth(), scr.getHeight());
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassViewEventHandler.java:678]: To correctly handle the case of a window straddling two screens, you need the screen associated with the window, not the screen that the anchor point intersects.

**Review by kevinrushforth — COMMENTED** [review]
This is closer, but doesn't correctly handle the case where a window straddles two screens. In that case, you really do want the screen associated with the window (even though that might seem a bit counter-intuitive). That's what's done pretty consistently in the rest of the code.

To test this, configure two screens with different scale factors, and move the window partially onto one screen such that most of the window (and importantly the scale associated with that window) is on one screen, an d the text insertion point is on the other.

**Review by kevinrushforth — APPROVED** [review]
Looks good. I tested it on a dual-screen system with different screen scales and it all works as expected.

**Comment by kevinrushforth** [comment]
@arapte Can you be the second reviewer?

**Review by arapte — APPROVED** [review]
Looks good , tested on Windows 11

**Pull request closed without merging**